sw/host/opentitanlib/src/transport/errors.rs

```diff
@@ -0,0 +1,137 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+/// Contains all the errors that any method on the `Transport` trait could generate.  This
+/// struct is serializable, such that it can be transmitted across a network for instance as
+/// part of the session proxy functionality.
+#[derive(Error, Debug, Serialize, Deserialize)]
+pub enum TransportError {
+    #[error("USB device did not match")]
+    NoMatch,
+    #[error("Found no USB device")]
+    NoDevice,
+    #[error("Found multiple USB devices, use --serial")]
+    MultipleDevices,
+    #[error("USB error: {0}")]
+    UsbGenericError(String),
+    #[error("Error opening USB device: {0}")]
+    UsbOpenError(String),
+    #[error("Transport does not support {0} instance {1}")]
+    InvalidInstance(TransportInterfaceType, String),
+    #[error("Encountered non-unicode device path")]
+    UnicodePathError,
+    #[error("Error opening {0}: {1}")]
+    OpenError(String, String),
+    #[error("Error reading from {0}: {1}")]
+    ReadError(String, String),
+    #[error("FPGA programming failed: {0}")]
+    FpgaProgramFailed(String),
+    #[error("Invalid pin strapping name \"{0}\"")]
+    InvalidStrappingName(String),
+    #[error("Transport does not support the requested operation")]
+    UnsupportedOperation,
+    #[error("Error communicating with FTDI: {0}")]
+    FtdiError(String),
+    #[error("Error communicating with debugger: {0}")]
+    CommunicationError(String),
+
+    // Include sub-enums for the various sub-traits of Tranport.
+    #[error("GPIO error: {0}")]
+    GpioError(crate::io::gpio::GpioError),
+    #[error("UART error: {0}")]
+    UartError(crate::io::uart::UartError),
+    #[error("SPI error: {0}")]
+    SpiError(crate::io::spi::SpiError),
+    #[error("SPI error: {0}")]
+    I2cError(crate::io::i2c::I2cError),
+}
+
+/// Enum value used by `TransportError::InvalidInstance`.
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+pub enum TransportInterfaceType {
+    Gpio,
+    Uart,
+    Spi,
+    I2c,
+}
+
+impl std::fmt::Display for TransportInterfaceType {
```

Comment:
Is there a specific reason these need to be uppercase, i.e. you can't defer to `Debug::fmt` here?

Reply:
There is no particular reason, but it had not occurred to me that you could use :? in the annotations of the error enums. Done.